### ISSUE TITLE
SpeakUp console-mode screen-reader causes kernel oops in vchiq_util

### ISSUE THREAD

**Issue by cromarty** [raspberrypi/linux#305]
Running SpeakUp console-mode screen-reader causes a kernel oops.  Sometimes quickly.  Best way to re-create oops is to press numpad+f1 to open help and scroll up and down help lines rapidly till oops occurs.

The stack trace:

```
pi@raspberrypi:~$ [  950.807592] Unable to handle kernel NULL pointer dereference at virtual address 00000008
[  950.817845] pgd = c0004000
[  950.821260] [00000008] *pgd=00000000
[  950.825724] Internal error: Oops: 5 [#1] PREEMPT ARM

Entering kdb (current=0xda9481c0, pid 28) Oops: (null)
due to oops @ 0xc0250584

Pid: 28, comm:              VCHIQ-0
CPU: 0    Tainted: G         C    (3.6.11+ #408)
PC is at vchiu_queue_is_empty+0x0/0x18
LR is at vchi_msg_dequeue+0x34/0xa0
pc : [<c0250584>]    lr : [<c024ff38>]    psr: 60000013
sp : da993f00  ip : 00000000  fp : c05c1ed0
r10: db000194  r9 : 00000014  r8 : da993f24
r7 : 00000014  r6 : da993f20  r5 : 00000000  r4 : 00000004
r3 : da993f20  r2 : 00000014  r1 : da993f24  r0 : 00000004
Flags: nZCv  IRQs on  FIQs on  Mode SVC_32  ISA ARM  Segment kernel
Control: 00c5387d  Table: 19ca4008  DAC: 00000017
[<c0013a7c>] (unwind_backtrace+0x0/0xf0) from [<c0072a84>] (kdb_dumpregs+0x28/0x50)
[<c0072a84>] (kdb_dumpregs+0x28/0x50) from [<c0074e08>] (kdb_main_loop+0x3a8/0x6fc)
[<c0074e08>] (kdb_main_loop+0x3a8/0x6fc) from [<c00774ec>] (kdb_stub+0x154/0x380)
[<c00774ec>] (kdb_stub+0x154/0x380) from [<c006e620>] (kgdb_handle_exception+0x1f8/0x668)
[<c006e620>] (kgdb_handle_exception+0x1f8/0x668) from [<c00131fc>] (kgdb_notify+0x24/0x40)
[<c00131fc>] (kgdb_notify+0x24/0x40) from [<c039b6d4>] (notifier_call_chain+0x44/0x84)
[<c039b6d4>] (notifier_call_chain+0x44/0x84) from [<c039b74c>] (__atomic_notifier_call_chain+0x38/0x4c)
[<c039b74c>] (__atomic_notifier_call_chain+0x38/0x4c) from [<c039b778>] (atomic_notifier_call_chain+0x18/0x20)
[<c039b778>] (atomic_notifier_call_chain+0x18/0x20) from [<c039b7b8>] (notify_die+0x38/0x44)
[<c039b7b8>] (notify_die+0x38/0x44) from [<c001102c>] (die+0xc4/0x39c)
[<c001102c>] (die+0xc4/0x39c) from [<c0394510>] (__do_kernel_fault.part.9+0x54/0x74)
[<c0394510>] (__do_kernel_fault.part.9+0x54/0x74) from [<c039b3e4>] (do_page_fault+0x1e8/0x3ec)
[<c039b3e4>] (do_page_fault+0x1e8/0x3ec) from [<c000832c>] (do_DataAbort+0x34/0x98)
[<c000832c>] (do_DataAbort+0x34/0x98) from [<c0399bd8>] (__dabt_svc+0x38/0x60)
Exception stack(0xda993eb8 to 0xda993f00)
3ea0:                                                       00000004 da993f24
3ec0: 00000014 da993f20 00000004 00000000 da993f20 00000014 da993f24 00000014
3ee0: db000194 c05c1ed0 00000000 da993f00 c024ff38 c0250584 60000013 ffffffff
[<c0399bd8>] (__dabt_svc+0x38/0x60) from [<c0250584>] (vchiu_queue_is_empty+0x0/0x18)
[<c0250584>] (vchiu_queue_is_empty+0x0/0x18) from [<c024ff38>] (vchi_msg_dequeue+0x34/0xa0)
[<c024ff38>] (vchi_msg_dequeue+0x34/0xa0) from [<bf074a94>] (audio_vchi_callback+0x48/0xac [snd_bcm2835])
[<bf074a94>] (audio_vchi_callback+0x48/0xac [snd_bcm2835]) from [<c024fd98>] (shim_callback+0x6c/0x138)
[<c024fd98>] (shim_callback+0x6c/0x138) from [<c0249b4c>] (slot_handler_func+0xdec/0x1714)
[<c0249b4c>] (slot_handler_func+0xdec/0x1714) from [<c003a7b4>] (kthread+0x88/0x94)
[<c003a7b4>] (kthread+0x88/0x94) from [<c000e9fc>] (kernel_thread_exit+0x0/0x8)
```

**Comment by popcornmix** [comment]
Has this ever worked reliably?

How do I install speakup? Do I need to build from git source, or is there a package?

**Comment by cromarty** [comment]
Hello,

No, SpeakUp has never yet worked reliably on the Raspberry Pi.

It turns out there is a fundamental reason for this.  The Broadcom GPU 
(Graphics Processing Unit) which processes both video and audio, despite 
it's name, is supported by the bcm2835 driver, which of course runs in 
kernel-space.

So there is a process which runs in user-space and which allows audio 
applications to interface to the chip by a message queueing system. 
This is the vchiq, or video core hardware interface queue.

SpeakUp regularly causes a kernel oops in this application by trying to 
dereference a null pointer.

About a month ago a change was made to the bcm2835 driver to add true 
mmap support.  This change SHOULD, remove the need for vchiq.  But it 
has introduced a nasty stutter in tts.

I have bugged both of these issues on the RPI foundation's issue 
tracker.  But there are currently almost 400 issues outstanding on the list.

I am currently trying to interface SpeakUp to USB audio, which will take 
vchiq out of the picture and should allow SpeakUp to run with stability.

If you are interested in the Pi take a look at our web site and maybe 
join our list:

http://www.raspberryvi.org/

List please excuse the advert.

Mike

On 28/05/2013 23:10, popcornmix wrote:

> Has this ever worked reliably?
> 
> How di I install speakup? Do I need to build from git source, or is
> there a package?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/raspberrypi/linux/issues/305#issuecomment-18583707.

## 

Michael A. Ray
Analyst/Programmer
Witley, Surrey, South-east UK

Ham Radio Call-sign: G4XBF, licensed since 1982

'Entia non sunt multiplicanda praeter necessitatem' - Ockham's razor

Raspberry VI:
http://www.raspberryvi.org/
NVDA, the best free screen-reader in the world:
http://www.nvda-project.org/

**Comment by licaon-kter** [comment]
Audio output via jack or hdmi?

**Comment by cromarty** [comment]
I've only tested it via 3.5mm audio jack.  As both analogue and HDMI 
audio uses vchiq I assume both.
Mike

On 29/05/2013 00:49, licaon-kter wrote:

> Audio output via jack or hdmi?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/raspberrypi/linux/issues/305#issuecomment-18587699.

## 

Michael A. Ray
Analyst/Programmer
Witley, Surrey, South-east UK

Ham Radio Call-sign: G4XBF, licensed since 1982

'Entia non sunt multiplicanda praeter necessitatem' - Ockham's razor

Raspberry VI:
http://www.raspberryvi.org/
NVDA, the best free screen-reader in the world:
http://www.nvda-project.org/

**Comment by skiphansen** [comment]
I'm experiencing this issue with another audio application I'm developing.  It occurs on 3.10.25+ #622 PREEMPT Fri Jan 3 18:41:00 GMT 2014 armv6l GNU/Linux.  It is not easy to reproduce, sometimes I can play audio for hours without an issue.  I've also only used the 3.5mm audio jack.

**Comment by cromarty** [comment]
To answer popcornmix's question about how to get speakup running, very late :) :+1: 

apt-get install espeak espeakup
update-rc.d espeakup defaults
modprobe speakup_soft

Bingo, speakup should be running, and stuttering like billyo.

**Comment by Ruffio** [comment]
@cromarty has this issue been resolved? If yes, then please close this issue.

**Comment by cromarty** [comment]
Hello.

No.  This is still the case.

It makes console speech with the SpeakUp screen-reader stutter very
badly, and with the Orca GUI screen-reader via speech-dispatcher the
same happens.

I wrote an OMX library for use with SpeakUp but not with speech-dispatcher.

Mike aka Cromarty

On 28/07/2016 09:53, Rasmus Christiansen wrote:

> @cromarty has this issue been resolved? If yes, then please close this issue.
> 
> ---
> 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/raspberrypi/linux/issues/305#issuecomment-235837889

## 

Michael A. Ray
Analyst/Programmer
Witley, Surrey, South-east UK

Eyes Free Linux:
http://eyesfreelinux.ninja/

Raspberry VI:
http://www.raspberryvi.org/

**Comment by JamesH65** [comment]
@cromarty I presume this issue is still present? If not please close.

**Comment by JamesH65** [comment]
This issue will be closed within 30 days unless further interactions are posted. If you wish this issue to remain open, please add a comment. A closed issue may be reopened if requested.

**Comment by cromarty** [comment]
I can confirm that I recently configured Raspbian to run SpeakUp via
espeakup and espeak in the way I did before, and now there is no
stuttering and no kernel oops.

Mike


On 27/06/2018 16:28, James Hughes wrote:
> This issue will be closed within 30 days unless further interactions are posted. If you wish this issue to remain open, please add a comment. A closed issue may be reopened if requested.
> 


-- 
Michael A. Ray
Analyst/Programmer
Witley, Surrey, South-east UK

"Perfection is achieved, not when there is nothing more to add, but when
there is nothing left to take away." -- A. de Saint-Exupery

https://cromarty.github.io/
http://eyesfreelinux.ninja/
http://www.raspberryvi.org/
